ipv8/peerdiscovery/network.py

```diff
@@ -1,141 +1,175 @@
 import logging
-from collections import OrderedDict
+from collections import OrderedDict, namedtuple
 from threading import RLock
+from typing import Dict, Iterable, List, Optional, Set
 
 from ..messaging.serialization import default_serializer
+from ..types import Address, Overlay, Peer
+
+
+MID = bytes
+PublicKeyMat = bytes
+Service = bytes
+ServiceSet = Set[Service]
+WalkableAddress = namedtuple('WalkableAddress', ['introduced_by', 'services', 'new_style'])
 
 
 class Network(object):
 
     def __init__(self):
-        # All known IP:port addresses, mapped to (introduction peer, services, new_style)
-        self._all_addresses = {}
-        # All verified Peer objects (Peer.address must be in _all_addresses)
-        self.verified_peers = set()
+        self._all_addresses: Dict[Address, WalkableAddress] = {}
+        '''All known IP:port addresses, mapped to (introduction peer, services, new_style)'''
+
+        self.verified_peers: Set[Peer] = set()
+        '''All verified Peer objects (Peer.address must be in _all_addresses)'''
+
+        self.verified_by_public_key_bin: Dict[PublicKeyMat, Peer] = {}
+        '''Map of known public keys to Peer objects.'''
+
         self.graph_lock = RLock()
-        # Peers we should not add to the network
-        # For example, bootstrap peers
-        self.blacklist = []
-        # Excluded mids
-        self.blacklist_mids = []
-
-        # Map of advertised services (set) per peer
-        self.services_per_peer = {}
-        # Map of service identifiers to local overlays
-        self.service_overlays = {}
-
-        # Cache of IP:port -> Peer
-        self.reverse_ip_lookup = OrderedDict()
+        '''Lock for all updates of the peer pool.'''
+
+        self.blacklist: List[Address] = []
+        '''Peers we should not add to the network (e.g., bootstrap peers), by address.'''
+
+        self.blacklist_mids: List[MID] = []
+        '''Peers we should not add to the network (e.g., bootstrap peers), by mid.'''
+
+        self.services_per_peer: Dict[PublicKeyMat, ServiceSet] = {}
+        '''Map of advertised services (set) per peer'''
+
+        self.service_overlays: Dict[Service, Overlay] = {}
+        '''Map of service identifiers to local overlays'''
+
         self.reverse_ip_cache_size = 500
+        self.reverse_ip_lookup = OrderedDict()
+        '''Cache of IP:port -> Peer. This is not a normal dict, as the amount of addresses can grow infinitely!'''
```

Comment:
The comment looks a bit unclear. `OrderedDict` does not provide some infinite memory to allow infinite growth :)
Probably it means that the old entries will be removed from the dict using `popitem(False)` which are missing from the usual dict?

Reply:
Reworded, should be a bit more clear now.